In [ ]:
import matplotlib.pyplot as plt
import numpy as np
from sqlalchemy import create_engine

Problem: Is revenue decline primarily caused by seasonal purchasing patterns 
	or by a small group of high-value customers reducing their purchases? 

In [60]:
engine = create_engine("mysql+pymysql://root:Vito32sql@localhost:3306/sales")
transactions_df = pd.read_sql("SELECT * FROM transactions_clean", engine)
customers_df = pd.read_sql("SELECT * FROM customers", engine)
transactions_df.info(),customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148672 entries, 0 to 148671
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   product_code   148672 non-null  object 
 1   customer_code  148672 non-null  object 
 2   market_code    148672 non-null  object 
 3   order_date     148672 non-null  object 
 4   sales_qty      148672 non-null  int64  
 5   sales_amount   148672 non-null  float64
 6   currency       148672 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 7.9+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38 entries, 0 to 37
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_code  38 non-null     object
 1   custmer_name   38 non-null     object
 2   customer_type  38 non-null     object
dtypes: object(3)
memory usage: 1.0+ KB


(None, None)

In [61]:
df = transactions_df.merge(customers_df, how='inner', on='customer_code')
df['order_date'] = pd.to_datetime(df['order_date'])
df['order_qtr'] = df['order_date'].dt.quarter
df['order_year'] = df['order_date'].dt.year

- Q4 Revenue decline is concentrated among major customers.
    - Electricalsara Stores, Electricalslytical, Excel Stores, Premium Stores, Nixon, Info Stores, Control, Surge Stores, Acclaimed Stores, Forward Stores, Epic Stores, Nomad Stores, Electricalsocity, Modular, Atlas Stores, Leader

- Quantify decline amongst Customers

In [62]:
top_10_customers = df.groupby(['custmer_name'])['sales_amount'].sum().sort_values(ascending=False).index[:10]

In [63]:
# Yearly
yearly_grp = df.groupby(['order_year'])
yearly_stats = yearly_grp.agg(
    revenue = ('sales_amount','sum'),
    orders = ('custmer_name','count')
).sort_index()


In [64]:
mask = df['custmer_name'].isin(top_10_customers)
top_10_yearly_stats = df[mask].groupby('order_year').agg(
    top_10_revenue = ('sales_amount','sum'),
    top_10_orders = ('custmer_name','count')
).sort_index()

In [66]:
yearly_10_rev_stats = pd.concat([yearly_stats['revenue'], top_10_yearly_stats['top_10_revenue']], axis='columns')
yearly_10_rev_stats['revenue_contribution%'] = (yearly_10_rev_stats['top_10_revenue']/yearly_10_rev_stats['revenue'])*100
yearly_10_rev_stats

,revenue,top_10_revenue,revenue_contribution%
order_year,,,
2017,93701152.0,72072791.0,76.917721
2018,414308943.0,314473638.0,75.903174
2019,336452114.0,246804344.0,73.354969
2020,142235559.0,106923812.0,75.173756


- 2017 only has 3 months of last quarter: 
    - Top 10 contribute 77% (7.2/9.3Cr)
- 2018:
    - Top 10 contribute 76% (31.4/41.4Cr)
- 2019:
    - Top 10 contribute 73% (24.6/33.6Cr)
- Revenue is highly dependant on the Top 10 customer across every year,
- When Top 10 contribution goes down by even 3% (In 2019) business faces revenue loss
- 41.4Cr in 2018 to 33.6Cr in 2019  
- Clear yearly decline in Top 10 customers contribution

In [67]:
revenue_decline_18_19 = yearly_stats.loc[2018,'revenue'] - yearly_stats.loc[2019,'revenue'] 
top_10_contribution_decline = yearly_10_rev_stats.loc[2018,'top_10_revenue'] - yearly_10_rev_stats.loc[2019,'top_10_revenue']
(top_10_contribution_decline / revenue_decline_18_19)*100

np.float64(86.91503991255539)

86.91% of the revenue decline between 2018-2019 came from the Top 10 customers

In [88]:
# Quarterly
year_qtr_grp = df.groupby(['order_year','order_qtr'])
year_qtr_stats = year_qtr_grp.agg(
    revenue = ('sales_amount','sum'),
    orders = ('custmer_name','count')
).sort_index()

In [89]:
top_10_y_q_stats = df[mask].groupby(['order_year','order_qtr']).agg(
    top_10_revenue = ('sales_amount','sum'),
    top_10_orders = ('custmer_name','count')
).sort_index()

In [91]:
year_qtr_10_rev_stats = pd.concat([year_qtr_stats['revenue'], top_10_y_q_stats['top_10_revenue']], axis='columns')
year_qtr_10_rev_stats['revenue_contribution%'] = (year_qtr_10_rev_stats['top_10_revenue']/year_qtr_10_rev_stats['revenue'])*100
year_qtr_10_rev_stats

revenue  top_10_revenue  revenue_contribution%
order_year order_qtr                                                    
2017       4           93701152.0      72072791.0              76.917721
2018       1          115950460.0      87133849.0              75.147480
           2          103118619.0      77398306.0              75.057547
           3          105790694.0      82011203.0              77.522133
           4           89449170.0      67930280.0              75.942885
2019       1           86945674.0      63671054.0              73.230848
           2           81123680.0      58811858.0              72.496536
           3           92465927.0      66179404.0              71.571666
           4           75916833.0      58142028.0              76.586477
2020       1           78970662.0      60830796.0              77.029614
           2           63264897.0      46093016.0              72.857174

In [71]:
# Quarters YoY
# Q1
def quarterly_YoY(qtr):
    rev_dec = year_qtr_10_rev_stats.loc[(2018,qtr),'revenue'] - year_qtr_10_rev_stats.loc[(2019,qtr),'revenue']
    top10_contibution_dec = year_qtr_10_rev_stats.loc[(2018,qtr),'top_10_revenue'] - year_qtr_10_rev_stats.loc[(2019,qtr),'top_10_revenue']
    return (top10_contibution_dec/rev_dec)*100
print(quarterly_YoY(1))
print(quarterly_YoY(2))
print(quarterly_YoY(3))
print(quarterly_YoY(4))

80.89283954723886
84.50329414416653
118.81482805665571
72.33231037624913


- Quarterly YoY(2018-2019) decline caused by Top 10 customers:
    - Q1: 81%
    - Q2: 84.5% 
    - Q3: 119%
    - Q4: 72.3%
- Q3 being over 100% means, theres growth in non-Top 10 customer revenue

In [72]:
non_top10_2018_q3_revenue = year_qtr_10_rev_stats.loc[(2018,3),'revenue'] - year_qtr_10_rev_stats.loc[(2018,3),'top_10_revenue'] 
non_top10_2019_q3_revenue = year_qtr_10_rev_stats.loc[(2019,3),'revenue'] - year_qtr_10_rev_stats.loc[(2019,3),'top_10_revenue'] 
non_top10_2019_q3_revenue - non_top10_2018_q3_revenue

np.float64(2507032.0)

- Q3 saw growth of 25lakhs in non-Top 10 customer revenue

- What causes the revenue drop in Top 10 customers? 

In [76]:
yearly_10_orders_stats = pd.concat([yearly_stats['orders'],top_10_yearly_stats['top_10_orders']],axis='columns')
yearly_10_orders_stats['top_10_order%'] = (yearly_10_orders_stats['top_10_orders']/yearly_10_orders_stats['orders'])*100
yearly_10_orders_stats

,orders,top_10_orders,top_10_order%
order_year,,,
2017,14604,8730,59.778143
2018,60866,35263,57.935465
2019,51817,26696,51.519772
2020,21385,10691,49.992986


- Yearly decline pattern in number of orders from Top 10 Customers

In [79]:
# Top 10 Orders YoY decline
((yearly_10_orders_stats.loc[2019,'top_10_orders'] - yearly_10_orders_stats.loc[2018,'top_10_orders'])/yearly_10_orders_stats.loc[2018,'top_10_orders'])*100

np.float64(-24.29458639367042)

- 2018-2019: 24.3% decline in orders from Top 10. 

In [96]:
year_qtr_10_orders_stats = pd.concat([year_qtr_stats['orders'], top_10_y_q_stats['top_10_orders']], axis='columns')
year_qtr_10_orders_stats['order_contribution%'] = (year_qtr_10_orders_stats['top_10_orders']/year_qtr_10_orders_stats['orders'])*100
year_qtr_10_orders_stats

orders  top_10_orders  order_contribution%
order_year order_qtr                                            
2017       4           14604           8730            59.778143
2018       1           15342           8800            57.358884
           2           15797           9318            58.985883
           3           15493           9060            58.478022
           4           14234           8085            56.800618
2019       1           13827           7742            55.991900
           2           13447           7212            53.632781
           3           13035           6191            47.495205
           4           11508           5551            48.236010
2020       1           11493           5712            49.699817
           2            9892           4979            50.333603

- Quarterly decline pattern in Top 10 orders from Q2-2018 to Q4-2019
    - Not considering 2017 and 2020 due to partial data
- Q4 of both years recorded least orders by Top 10 customers.
    - 2018-Q4: 8085 orders
    - 2019-Q4: 5551 orders

- Risk assessment of each customer in Top 10

In [ ]:
business_revenue = df['sales_amount'].sum()
def customer_revenue_share(df,customer):
    mask = (df['custmer_name'] == customer)
    customer_revenue = df.loc[mask]['sales_amount'].sum()
    revenue_share = ((customer_revenue/business_revenue)*100)
    return round(revenue_share,1)

In [159]:
def customer_decline(df,customer):
    mask = (df['custmer_name'] == customer)
    cond1 = (df['order_year'] == 2018)
    cond2 = (df['order_year'] == 2019)
    decline_amount = (df.loc[mask & cond2]['sales_amount'].sum()) - (df.loc[mask & cond1]['sales_amount'].sum())
    return round(decline_amount,2)

In [167]:
risk_assessment = []
for customer in top_10_customers:
    revenue_share = customer_revenue_share(df,customer)
    revenue_change = customer_decline(df,customer)
    record = {
        'customer' : customer,
        'revenue share' : revenue_share,
        'revenue change' : revenue_change
    }
    if (revenue_share > 3) & (revenue_change < -1000000):
        record['risk'] = 'Critical'     # High share and declining
    elif (revenue_share > 3) & (revenue_change < 0):
        record['risk'] = 'Safe'         # High share and stable
    elif (revenue_share > 0) & (revenue_change < 0):
        record['risk'] = 'Monitor'      # Low shares and declining
    elif (revenue_change > 0):
        record['risk'] = 'Growth'       # growing
    risk_assessment.append(record)
risk_df = pd.DataFrame(risk_assessment)

In [178]:
critical_decline = (risk_df[risk_df['risk'] == 'Critical']['revenue change'].sum()*-1)
(critical_decline/revenue_decline_18_19)*100 

np.float64(77.03275842379864)

- 2018-2019: 77% Revenue decline is caused by the critical customers amongst top 10
- Seasonality exists, but the business's revenue decline is overwhelmingly driven by concentration risk and declining purchases from a handful of major customers.

##### Conclusion — What is driving the revenue decline, and what should the business prioritise to reverse it?

The analysis answers the core question directly: **the revenue decline is a customer concentration problem, not a seasonal one.** The top 10 customers account for 73–77% of total revenue in every year on record, and 86.9% of the 2018→2019 decline of ₹7.8 Cr is attributable to them alone. Seasonality exists as a background pattern but is not the causal driver.

**Priority actions:**

1. Immediate retention effort on Critical-tier customers. Four customers combine a revenue share above 3% with a year-on-year decline exceeding ₹10L — together they account for 77% of the total decline. These are known by name. Account-level conversations, commercial reviews, or pricing adjustments should happen before any other initiative.

2. Investigate the order-volume drop separately from the revenue drop. Top 10 orders fell 24.3% between 2018 and 2019. Whether this reflects customers switching to competitors, reducing operations, or renegotiating terms is not answerable from transaction data alone — but the distinction determines the right response.

3. Protect and develop the non-Top 10 base. Q3 2019 is the one bright spot in the data: non-Top 10 revenue grew by ₹25L even as the top tier contracted. This segment is currently absorbing some of the concentration risk and deserves deliberate investment rather than benign neglect.

**What the data ruled out:**

- Seasonality is not the primary cause. Q4 records the lowest order volumes in both 2018 and 2019, but the decline is consistent across all four quarters — Q1 through Q4 YoY declines are explained 72–119% by Top 10 customer behaviour, not by time-of-year patterns.
- Broad demand weakness is not the cause. The non-Top 10 segment held steady or grew. The problem is concentrated, not systemic.

**Data limitations to note:**

- The dataset covers 2017–2019, with 2017 containing only Q4 and 2020 presumably partial — multi-year trend lines should be read with this in mind.
- There is no visibility into *why* top customers reduced orders: no CRM data, no contract terms, no competitor context. The risk framework (Critical / Safe / Monitor / Growth) correctly flags who to act on, but the business will need qualitative investigation to understand what changed.
- Customer-level revenue share is computed across the full dataset period, which dilutes the 2019-specific risk signal slightly — a customer declining sharply in 2019 may appear less critical if their 2017–2018 share was high.

visualisations built in Power BI dashboard